### ISSUE TITLE
2-12 seconds of UI hang (not sure if it's Sembast)

### ISSUE THREAD

**Issue by tobinharris** [tekartik/sembast.dart#250]
Hi

Thanks for making Sembast, it's made part of our development really straight forward.

We have an app in Beta test and 51% of our users are reporting 2-10 seconds of UI hang (iOS and Droid). We cannot recreate on our development/test devices or simulators. Nor on a device farm. But we've seen videos of it.

Our Sembast DB:
 
* About 18MB
* SQLite store
* Biggest `value` of an `Entry` is 10Kb
* 3,000-4,000 `Entry` records in SQLite db

Data Access Patterns

* 3,000 records downloaded and inserted on first use in a single tx (we onboard the user whilst this happens)
* Users can add their own records (usually only 10-20 a day)
* They will do searches records, which is where the biggest `reads` happen
* We re-fetch the data at most once per 24 hours, in the background. Usually takes a minute on a good connection.

When it hangs

* Seems to be anywhere in the app. Any screen. Not just after writes or reads.
* Can be 1 second, and at worst up to 15 seconds.
* It's different to Jank in that the UI actually locks up and won't respond.

Could any of this be related to Sembast do you think?

**Comment by alextekartik** [comment]
Thanks for the report. In theory, sembast should handle this with its weird concept of a 'cooperator' (I was wondering why nobody complain about that so far, here it is!). Since sembast runs in the main isolate, heavy algorithm are paused (every 16 millis) to let other tasks (including UI) run. I was able to test with 100k record on an old nexus successfully (basic list with edit and sort).

Now that I review the code on this (cooperate()), I can see that if you a lot of concurrent sembast task (by task here I mean calls), the implementation is wrong. Too bad that you cannot reproduce.

I will make some more experiments on this today or tomorrow and will keep you posted. As a side note, I think sembast should handle gracefully your database requirements.

Some questions.
- Which version of sembast are you using? (nnbd?). 
- Do you have a lot of listeners (onSnapshot) and do you cancel them properly?
- Does the hang happens right away of after some time in the app?
- Do you read 3K entry and convert them (such as fromJson to an object) in the main isolate (basically do you have heavy algorithm (sort and conversion for example) on a list of snapshots?
- If I could see and run your source code to get into a reproducible scenario, that would be great (but I understand that it is not always feasible).

**Comment by tobinharris** [comment]
Hi Alex

Huge thanks for the quick reply. 

> Which version of sembast are you using? (nnbd?).

    sembast-3.0.0+4
    sembast_sqflite-2.0.0+1

> Do you have a lot of listeners (onSnapshot) and do you cancel them properly?

None.

> Does the hang happens right away of after some time in the app?

Can be pretty quick. Some examples:

    Home Screen --> Pressed to Navigate --> Food Diary --> Swiped to delete an item --> Item deleted and 'Undo' snackbar appeared. 
    The screen froze and I could not scroll or delete another item for around 7 seconds.

    Was on Food Diary --> Pressed the back button to get to home screen --> Pressed Restaurants and Categories. 
    Freeze for around 2-3 seconds.

> Do you read 3K entry and convert them (such as fromJson to an object) in the main isolate (basically do you have heavy algorithm (sort and conversion for example) on a list of snapshots?

Yes, the search screen is fairly intense. It's a custom query, sort, then map to objects. However, the freeze happens in other places where this algorithm isn't used. We do map JSON to objects pretty much everywhere though, usually only 100-200 entries on any view.

> If I could see and run your source code to get into a reproducible scenario, that would be great (but I understand that it is not always feasible).

This might be possible, and thank you for offering. We may need an NDA but would happily sponsor you (or buy your cool guitar apps!).

**Comment by alextekartik** [comment]
Can you try to update sembast to 3.0.2? Basically previously there was a risk if multiple heavy queries were in progress that the 'cooperator' would not pause properly.

After that the next step I could think of would be to put a stopwatch around some suspicious request/algorithm in your code (by suspicious I mean the one where you think it could take some time) to find the bad guy which is a painful task.

As a side note, using listeners (onSnapshot) on a query is a good way to avoid performing unneeded refresh and is very efficient if you have a big list (and it works well with flutter)

**Comment by tobinharris** [comment]
Cheers @alextekartik. Is 3.0.2 released, we couldn't see it? It would definitely be worth a try.

Appreciate the other suggestions too.

**Comment by alextekartik** [comment]
yeap it is. pub upgrade or forced dependency (`sembast: '>=3.0.2'`) should work.

**Comment by tobinharris** [comment]
I just realised, we're only including `sembast_sqflite: ^2.0.0+1` in our `pubspec.yaml`. Is there an updated version of this to your knowledge?

**Comment by alextekartik** [comment]
sembast_sqflite 2.0.0+1 is the latest so far (re-publishing just to change the min version of a dep is not something many people do=. Yes you don't have add sembast as a dependency, here the goal was just to force the update. But as I said, `pub upgrade` should do the job!

**Comment by tobinharris** [comment]
Don't want to speak to soon, but we're running a test now on iOS and Android so we'll see if the above recommendation has fixed the issue. Thanks for the help so far.

**Comment by tobinharris** [comment]
HI @alextekartik 

Thanks for your help so far, we think that `sembast: '>=3.0.2` may have improved thing. This time our BETA test resulted in 25% of people experiencing "freeze" problems rather than 51%. 

One user reported that it only started freezing after a few days (on iOS). I'm wondering if it could be cumulative, but then again we're not sure iOS would keep an app alive for a few days anyway.

With your idea of a stopwatch, since we can't duplicate any freeze on our development devices, do you think we should log  the results and send to Beta testers? I'm trying to think how we can catch the freeze in the wild?

**Comment by alextekartik** [comment]
The stopwatch idea was more something to do on development devices as it might involve trying it at many different places. I'm not aware of any good solution for release reporting. I would focus on trying to reproduce the issue first but I know that sometimes it is not something that can easily be done. Sorry for not having better ideas.

**Comment by tobinharris** [comment]
Just a quick update. We've spotted a large GC event coincide with the 5-30 second freeze. We suspect this is more likely to be because images are loaded for each record (the bytes allocated is about 800K per thumbnail). I did noticed that there are thousands of Sembast store references allocated too, is this normal?

![image](https://user-images.githubusercontent.com/25578/117611922-3c32c580-b15c-11eb-9896-2099371cd445.png)
